crates/ty_python_semantic/resources/mdtest/assignment/annotations.md

```diff
@@ -234,3 +234,48 @@ reveal_type(x)  # revealed: Foo
 x: int = 1
 reveal_type(x) # revealed: Literal[1]
 ```
+
+## Annotations influence generic call inference
+
+```toml
+[environment]
+python-version = "3.12"
+```
+
+```py
+from typing import Literal
+
+def f[T](x: T) -> list[T]:
+    return [x]
+
+a = f("a")
+reveal_type(a)  # revealed: list[Literal["a"]]
+
+b: list[int | Literal["a"]] = f("a")
+reveal_type(b)  # revealed: list[int | Literal["a"]]
+
+c: list[int | str] = f("a")
+reveal_type(c)  # revealed: list[int | str]
+
+d: list[int | tuple[int, int]] = f((1, 2))
+reveal_type(d)  # revealed: list[int | tuple[int, int]]
+
+e: list[int] = f(True)
+reveal_type(e)  # revealed: list[int]
+
+# TODO: the RHS should be inferred as `list[Literal["a"]]` here
+# error: [invalid-assignment] "Object of type `list[int | Literal["a"]]` is not assignable to `list[int]`"
+g: list[int] = f("a")
```

Comment:
Ideally we would infer the RHS without the type context if the unannotated RHS is not assignable to the type annotation. However, this is a little difficult as we also want the type annotation to be the first type that we use for inference, so that the inferred type more closely matches the type annotation.

I think the only way this can work is if we perform inference twice, first without the type context, and then again with it only if the unannotated inferred type is assignable to the type annotation. However, I'm not sure that's worth the extra work for this specific error message.

Reply:
Yeah, I think what would make sense here is to re-infer the RHS without context if we are emitting a diagnostic. Some extra cost is fine, if it's only paid when actually emitting a diagnostic.